lib/attrsets.nix

```diff
@@ -333,6 +333,39 @@ rec {
       ) (attrNames set)
     );
 
+   /*
+    Like builtins.foldl' but for Attrsets.
+    Iterates over every nameValuePair in the given AttrSet.
+    The result can be appended to `acc` which gets finally returned.
+    Example
+      reduceAttrs
+      (acc: {name, value}:
+        acc // {
+          sum = acc.sum or 0 + value;
+          names = acc.names or [] ++ [name];
```

Comment:
`++` is `O(n)`, which in this context produces `O(n²)` behavior. Not great, although I think we could perhaps optimize the interpreter to detect chains of list concatenation thunks and process them efficiently. Would have to get back on this.